scipy/sparse/_coo.py

```diff
@@ -286,13 +293,17 @@ def toarray(self, order=None, out=None):
         fortran = int(B.flags.f_contiguous)
         if not fortran and not B.flags.c_contiguous:
             raise ValueError("Output array must be C or F contiguous")
-        if self.ndim > 2:
-            raise ValueError("Cannot densify higher-rank sparse array")
         # This handles both 0D and 1D cases correctly regardless of the
         # original shape.
-        M, N = self._shape_as_2d
-        coo_todense(M, N, self.nnz, self.row, self.col, self.data,
-                    B.ravel('A'), fortran)
+        if self.ndim < 3:
+            M, N = self._shape_as_2d
+            coo_todense(M, N, self.nnz, self.row, self.col, self.data,
+                        B.ravel('A'), fortran)
+            
+        else:
+            coords = np.concatenate(self.coords)
+            coo_todense_nd(self.shape, self.nnz, self.ndim,
+                           coords, self.data, B.ravel('A'), fortran)
```

Comment:
I thought you found that the numpy method was better than the sparsetools for nd. Can you explain what your plans are for sparsetools vs numpy tools? I think we talked about it in a zoom call, but it'd be nice to have it in github comments -- and I'm not sure I understood correct the approach you plan to take.

Reply:
Hi! From what I recall, the approach we all seemed to like was that we will initially try to get full support for n-D sparse arrays in at least one format. Once we have that fully working, if time permits we could then go back and look into optimising / refactoring the code. We will only make minimal C++ additions, and only those that are truly necessary (for which the NumPy alternatives are too slow).

As for the sparsetools vs numpy tools plans -

- sparse-dense add uses `coo_todense_nd` from sparsetools.
- sparse-sparse add and sub use numpy tools.
- mat-vec multiplication (n-D sparse * 1-D dense) uses `coo_matvec_nd` from sparsetools.

For the methods that will possibly be in the follow-up PR -

- `tensordot` and `dot` are a combination of numpy tools and pre-existing CSR sparsetools. They are mapping two n-D sparse arrays to two 2-D ones, using 2-D CSR sparse-sparse multiplication, and mapping the product back to n-D. I've tested this approach and it seems to work fine.
- sparse-sparse `matmul` is currently work in progress. The approach in mind is to convert n-D COO arrays to block diagonal form (i.e. a 2-D COO array), use CSR matmul, then convert the product back to n-D.

We are yet to look into what will be more efficient for sparse-dense `dot`, `tensordot`, and `matmul`, (we might need to work with compiled code like we did for sparse-dense matvec and sparse-dense addition) but we are aiming to get the sparse-sparse operations for the three working properly first.

But, again, all of the above will be in a follow-up PR. This PR is necessary as it makes the required changes to allow n-D indices in the COO data structure.